odxtools/diagcomm.py

```diff
@@ -0,0 +1,241 @@
+
+        functional_class_refs = [
+            odxrequire(OdxLinkRef.from_et(el, doc_frags))
+            for el in et_element.iterfind("FUNCT-CLASS-REFS/FUNCT-CLASS-REF")
+        ]
+
+        audience = None
+        if (audience_elem := et_element.find("AUDIENCE")) is not None:
+            audience = Audience.from_et(audience_elem, doc_frags)
+
+        protocol_snrefs = [
+            odxrequire(el.get("SHORT-NAME"))
+            for el in et_element.iterfind("PROTOCOL-SNREFS/PROTOCOL-SNREF")
+        ]
+
+        related_diag_comm_refs_raw = [
+            RelatedDiagCommRef.from_et(el, doc_frags)
+            for el in et_element.iterfind("RELATED-DIAG-COMM-REFS/RELATED-DIAG-COMM-REF")
+        ]
+
+        pre_condition_state_refs = [
+            odxrequire(OdxLinkRef.from_et(el, doc_frags))
+            for el in et_element.iterfind("PRE-CONDITION-STATE-REFS/PRE-CONDITION-STATE-REF")
+        ]
+
+        state_transition_refs = [
+            odxrequire(OdxLinkRef.from_et(el, doc_frags))
+            for el in et_element.iterfind("STATE-TRANSITION-REFS/STATE-TRANSITION-REF")
+        ]
+
+        semantic = et_element.get("SEMANTIC")
+
+        diagnostic_class: Optional[DiagClassType] = None
+        if (diagnostic_class_str := et_element.get("DIAGNOSTIC-CLASS")) is not None:
+            try:
+                diagnostic_class = DiagClassType(diagnostic_class_str)
+            except ValueError:
+                diagnostic_class = cast(DiagClassType, None)
+                odxraise(f"Encountered unknown diagnostic class type '{diagnostic_class_str}'")
+
+        is_mandatory_raw = odxstr_to_bool(et_element.get("IS-MANDATORY"))
+        is_executable_raw = odxstr_to_bool(et_element.get("IS-MANDATORY"))
+        is_final_raw = odxstr_to_bool(et_element.get("IS-FINAL"))
+
+        return DiagComm(
+            admin_data=admin_data,
+            sdgs=sdgs,
+            functional_class_refs=functional_class_refs,
+            audience=audience,
+            protocol_snrefs=protocol_snrefs,
+            related_diag_comm_refs_raw=related_diag_comm_refs_raw,
+            pre_condition_state_refs=pre_condition_state_refs,
+            state_transition_refs=state_transition_refs,
+            semantic=semantic,
+            diagnostic_class=diagnostic_class,
+            is_mandatory_raw=is_mandatory_raw,
+            is_executable_raw=is_executable_raw,
+            is_final_raw=is_final_raw,
+            **kwargs)
+
+    @property
+    def functional_classes(self) -> NamedItemList[FunctionalClass]:
+        return self._functional_classes
+
+    @property
+    def protocols(self) -> NamedItemList["DiagLayer"]:
+        return self._protocols
+
+    @property
+    def related_diag_comm_refs(self) -> NamedItemList[RelatedDiagCommRef]:
+        return self._related_diag_comm_refs
+
+    @property
+    def pre_condition_states(self) -> NamedItemList[State]:
+        return self._pre_condition_states
```

Comment:
why the property if there is no logic?

Reply:
to make this attribute unsettable (by normal means)

Reply:
why? most attributes in other classes are settable

Reply:
only the parts that can be defined using the XML. computed stuff like resolved references is not supposed to be user settable. (there are plenty examples: [1](https://github.com/mercedes-benz/odxtools/blob/main/odxtools/tablerow.py#L140), [2](https://github.com/mercedes-benz/odxtools/blob/main/odxtools/diaglayer.py#L350), [3](https://github.com/mercedes-benz/odxtools/blob/main/odxtools/odxtools/database.py#L147), ...)